corehq/sql_db/shard_data_management.py

```diff
@@ -99,6 +99,56 @@ def get_count_of_unmatched_models_by_shard(database, model):
     return results
 
 
+def delete_unmatched_shard_data(database, model, query_size=5000):
+    """
+    Deletes any `model` data residing on shards that aren't expected to live in `database`.
+
+    The deletes are batched in query_size and sorted by the model's partition_attr.
+
+    Returns a generator where each value is a tuple of last value included in the query and number of records deleted
+    """
+    column_name = model.partition_attr
+    shard_id_function = _get_shard_id_function(model)
+
+    # Intentionally uses BETWEEN because it is inclusive of both values
+    query = """
+    DELETE FROM {table_name}
+    WHERE NOT ({shard_id_function}) = ANY(%s) AND {column_name} BETWEEN %s AND %s;
+    """.format(
+        shard_id_function=shard_id_function,
+        table_name=model._meta.db_table,
+        column_name=column_name
+    )
+
+    valid_shards = partition_config.get_shards_on_db(database)
+    qs = model.objects.using(database).order_by(column_name).values_list(column_name, flat=True)
+    prev_value = None
+    value = qs.first()
+    filter_expression = {}
+
+    last_iteration = False
+    while not last_iteration:
+        qs = qs.filter(**filter_expression)
+        prev_value = value
+        try:
+            value = qs[query_size]
```

Comment:
Is `qs` supposed to be limited by `query_size` somewhere? I don't see where that's happening. Also, this should probably be changed if the queryset is expected to have a max length of `query_size`
```suggestion
            value = qs[query_size - 1]
```